### DISCUSSION TITLE
UI: Add Short Names for Commands (for 3rd party addon Ribbon WB)

### DISCUSSION THREAD

**Discussion by geolta** [FreeCAD/FreeCAD#16156] [Discussions that will turn in to issues but need refining]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

*Based on the discussion in https://forum.freecad.org/viewtopic.php?t=79235*

For a Ribbon UI, short names for commands are required since they are displayed below the button if it is a large button. For an entry in the menu bar, longer names are okay (e.g. "Create Polyline" for menu and just "Polyline" for the ribbon). Until now, the texts in `menuText` are optimized for classic menu bars and no additional short name is available.

![Using MenuText](https://github.com/FreeCAD/FreeCAD/assets/115890454/c7767765-1a17-4009-88ad-e6c2651ccf71)
vs. (although the margins are now maybe too small – but this is an issue not related to the text problem)
![Using shorter Text](https://github.com/FreeCAD/FreeCAD/assets/115890454/298c1162-5b92-4b42-bfbb-1f47376b7bd7)

For solving this problem, I see two potential solutions:
- add `shortName` or `ribbonName` property to command
- create additional file (XML, JSON, whatever) where these short names are saved
- more ideas?

Furthermore, the solution should fulfill the following requirements:
- translatable
- working for multiple FreeCAD versions
- the developer is more or less forced to at least think about a short name to avoid a new command is added without being compliant to a ribbon UI

In my opinion (but my experience with FreeCAD code base is quite limited) these requirements are pointing towards extending the `Command` object. What are you thinking about this?

*(I want to emphasize, that solving this problem is a prerequisite for implementing a good ribbon UI, regardless of implementing it in Python or C++.)*

### Full version info

```shell
OS: Fedora Linux 38 (KDE Plasma) (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.20.2.Unknown
Build type: Unknown
Python 3.11.1, Qt 5.15.8, Coin 4.0.0, Vtk 9.2.5, OCC 7.6.3
```

### Subproject(s) affected?

None

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ghost** [comment]
This does not belong here, it should moved to the ribbon ui add-on tracker.

**Comment by geolta** [comment]
I do think that it belongs here since the changes apply to FreeCAD master code, not to some addon.

**Comment by luzpaz** [comment]
I like the idea of `shortname`.

**Comment by ghost** [comment]
> I do think that it belongs here since the changes apply to FreeCAD master code, not to some addon.

So you want to change something in the master for an addon, make little sense to me.

**Comment by 80avin** [comment]
I second the idea of doing the change in Freecad than the plugin.
A few points in support of the issue which aren't mentioned previously could be:

Why such names are generally not good for UI elements ? :
* From the perspective of UI design, "Create" in "Create Polyline" is just redundant and wastes unnecessary space while creating more text for the user to scan. It would've been fine if there were only one such command, but when multiple commands are named "Create Polyline", "Create Rectangle", etc, this issue becomes more prominent.
* Particularly in the "Constraints" group ( and similar ), when the command is already defined to be a part of the "Constraints" group, it is completely unnecessary to prefix every command with "Constraint " text

Considering that such names are generally not good for UI elements, Why should this be fixed in FreeCAD and not the plugin ? :
* FreeCAD code is the source of this naming pattern, and if the pattern is wrong, it should be fixed at the source.
* Asking the UI plugin to have the "Create Polyline" to "Polyline" mapping means that
   * All such UI plugins need to have such mapping.
   * They require this mapping for all supported languages, which would slow down the plugin development or affect the quality of an entire class of such plugins.
   * Any new command added by any plugin or FreeCAD update can break the UI plugin which makes this issue an issue of inter-compatibility of multiple plugins, suggesting that it should be fixed in FreeCAD itself.

###### Feel free to correct me.

**Comment by maxwxyz** [comment]
I think instead of adding something for one addon we should discuss options of changing the UI in main.

**Reply by 80avin** [reply]
That is a much better idea.
However, I'll add that if "changing UI in main" requires adding shortnames to the menu commands or renaming the menu items from "Constrain Distance" to "Distance", we should do that first.

For 2nd, if someone wants to show the menu item outside of group, he can prepend group name to the option like "Constraint > Distance"

**Reply by pierreporte** [reply]
@kadet1090 will need this for his ribbon, and add-on developers will be able to make their own ribbon

**Comment by Reqrefusion** [comment]
The problem here, in my opinion, is there a need for a short name? Or rather, what is the use of a long name anymore? I think the way to go is to shorten existing names. When a name reaches a certain length, it becomes dysfunctional. The existence of verbs is not really necessary. At least for names.

**Reply by kadet1090** [reply]
Generally speaking from UX perspective you want to have things that are as clear and easy to understand as possible, from UI on the other hand you often fight with limited space and so want thing as concise as possible. Job of the designer is to properly balance these two. 

Here we have a problem of contextuallity. Using "Tangent" as label for action that is in "Constraint" group should be clear and easy to understand for the user - it will create "Tangent" constraint. Without the context of "Constraint" group the "Tangent" label could however be used also for other things. For example we can have "Tangent" labelled button in "Select Geometry" group - this time implied action is that it will select all tangent geometry. 

Without knowledge of context it is hard to provide good and easy to understand labels that is why the disambiguation of action is done via longer label "Create Tangent Constraint" is clear regardless of where it is placed.  As the commands are defined globally and not contextually we can't define a good short label in most cases, but there should be a way to provide different labels when context is known.

**Reply by Reqrefusion** [reply]
Names in your hypothetical universe are no longer names, they are descriptions. I think that name is synonymous with short name. Since I am not involved in this project, it is pointless to insist that my assumption is correct, because the other approach is not wrong. But there should be short names and comprehensive descriptions. Anything more will cause confusion and problems.
